pylabrobot/liquid_handling/backends/tecan/EVO.py

```diff
@@ -0,0 +1,740 @@
+"""
+This file defines interfaces for all supported Tecan liquid handling robots.
+"""
+
+# pylint: disable=invalid-name
+
+from abc import ABCMeta, abstractmethod
+from typing import Dict, List, Optional, Tuple, Sequence, TypeVar, Union
+
+from pylabrobot.liquid_handling.liquid_classes.tecan import TecanLiquidClass, get_liquid_class
+from pylabrobot.liquid_handling.backends.USBBackend import USBBackend
+from pylabrobot.liquid_handling.standard import (
+  Pickup,
+  PickupTipRack,
+  Drop,
+  DropTipRack,
+  Aspiration,
+  AspirationPlate,
+  Dispense,
+  DispensePlate,
+  Move
+)
+from pylabrobot.resources import (
+  TecanPlate,
+  TecanTipRack,
+  TecanTip
+)
+
+T = TypeVar("T")
+
+class TecanLiquidHandler(USBBackend, metaclass=ABCMeta):
+  """
+  Abstract base class for Tecan liquid handling robot backends.
+  """
+
+  @abstractmethod
+  def __init__(
+    self,
+    packet_read_timeout: int = 3,
+    read_timeout: int = 30,
+    write_timeout: int = 30,
+  ):
+    """
+
+    Args:
+      packet_read_timeout: The timeout for reading packets from the Tecan machine in seconds.
+      read_timeout: The timeout for reading from the Tecan machine in seconds.
+    """
+
+    super().__init__(
+      packet_read_timeout=packet_read_timeout,
+      read_timeout=read_timeout,
+      write_timeout=write_timeout,
+      id_vendor=0x0C47,
+      id_product=0x4000)
+
+    self._cache: Dict[str, List[Optional[int]]] = {}
+
+  def _assemble_command(
+    self,
+    module: str,
+    command: str,
+    params: List[Optional[int]]) -> str:
+    """ Assemble a firmware command to the Tecan machine.
+
+    Args:
+      module: 2 character module identifier (C5 for LiHa, ...)
+      command: 3 character command identifier
+      params: list of integer parameters
+
+    Returns:
+      A string containing the assembled command.
+    """
+
+    cmd = module + command + ",".join(str(a) if a is not None else "" for a in params)
+    return f"\02{cmd}\00"
+
+  def parse_response(self, resp: str) -> Dict[str, Union[str, List[int]]]:
+    """ Parse a machine response string
+
+    Args:
+      resp: The response string to parse.
+
+    Returns:
+      A dictionary containing the parsed values.
+    """
+
+    data: List[int] = [int(x) for x in resp[3:-1].split(",")]
+    return {
+      "module": resp[1:3],
+      "ret": [0], # data[0] TODO: get error return code
```

Comment:
This was the byte thing, right?

Reply:
Yup. I'll capture an example error in Wireshark when I use the machine again.